x/evm/ante/preprocess.go

```diff
@@ -39,35 +46,53 @@ func (p EVMPreprocessDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate
 		// this would never happen if this handler call is routed by the router
 		return ctx, errors.New("no message exists in EVM tx")
 	}
-	ethTx, txData := tx.GetMsgs()[0].(*evmtypes.MsgEVMTransaction).AsTransaction()
-	ctx = evmtypes.SetContextEthTx(ctx, ethTx)
+	msgEVMTransaction := tx.GetMsgs()[0].(*evmtypes.MsgEVMTransaction)
+	txData, err := evmtypes.UnpackTxData(msgEVMTransaction.Data)
+	if err != nil {
+		return ctx, err
+	}
 	ctx = evmtypes.SetContextTxData(ctx, txData)
 	// use infinite gas meter for EVM transaction because EVM handles gas checking from within
 	ctx = ctx.WithGasMeter(sdk.NewInfiniteGasMeter())
 
-	V, R, S := ethTx.RawSignatureValues()
 	chainID := p.evmKeeper.ChainID(ctx)
 	evmParams := p.evmKeeper.GetParams(ctx)
 	chainCfg := evmParams.GetChainConfig()
 	ethCfg := chainCfg.EthereumConfig(chainID)
 	ctx = evmtypes.SetContextEtCfg(ctx, ethCfg)
 	version := GetVersion(ctx, ethCfg)
+	signer := SignerMap[version](ethCfg.ChainID)
+	if atx, ok := txData.(*ethtx.AssociateTx); ok {
+		V, R, S := atx.GetRawSignatureValues()
+		V = new(big.Int).Add(V, big.NewInt(27))
+		evmAddr, seiAddr, _, err := getAddresses(V, R, S, common.Hash{}) // associate tx should sign over an empty hash
+		if err != nil {
+			return ctx, err
+		}
+		if _, found := p.evmKeeper.GetEVMAddress(ctx, seiAddr); found {
+			return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "account already has association set")
+		}
+		seiBalance := p.evmKeeper.BankKeeper().GetBalance(ctx, seiAddr, p.evmKeeper.GetBaseDenom(ctx)).Amount
+		evmBalance := new(big.Int).SetUint64(p.evmKeeper.GetBalance(ctx, evmAddr))
+		if new(big.Int).Add(seiBalance.BigInt(), evmBalance).Cmp(new(big.Int).SetUint64(BalanceThreshold)) < 0 {
+			return ctx, sdkerrors.Wrap(sdkerrors.ErrInsufficientFunds, "account needs to have at least 1Sei to force association")
+		}
+		p.evmKeeper.SetAddressMapping(ctx, seiAddr, evmAddr)
+		return ctx.WithPriority(math.MaxInt64 - 101), nil // short-circuit without calling next
```

Comment:
why do you subtract 101 from MaxInt64 here?

Reply:
we want it to be high priority but not higher than oracle's. I've refactored it into constants to be more descriptive